experiments/2017-09-21-rust-on-android.md

```diff
@@ -0,0 +1,224 @@
+    * LLDB 
+
+Once the NDK and associated tools have been installed, we need to set a few environment variables, first for the SDK path and the second for the NDK path. To find out where  these items are installed, go to File > Project Structure. Here you will find the paths to the SDK and NDK. Set the following envvars:
+```
+export ANDROID_HOME=<SDKPath>
+export NDK_HOME=<NDKPath>
+```
+
+If you do not already have Rust is installed, we need to do this now. For this we will be using [rustup](https://www.rustup.rs/). Rustup installs Rust from the official release channels and enables you to easily switch between different release versions. It will be useful to you for all your future Rust development, not just here.
+```
+curl https://sh.rustup.rs -sSf | sh
+```
+
+The next step is to create standalone versions of the NDK for us to compile against. We need to do this for each of the architectures we want to compile against. We will be using the `make_standalone_toolchain.py` script inside the main Android NDK in order to do this. First create a directory for our project.
+```
+mkdir greetings
+cd greetings
+```
+
+Now let's create our standalone NDKs.
+```
+mkdir NDK
+${NDK_HOME}/build/tools/make_standalone_toolchain.py --unified-headers --api 21 --arch arm64 --install-dir NDK/arm64
+${NDK_HOME}/build/tools/make_standalone_toolchain.py --unified-headers --api 14 --arch arm --install-dir NDK/arm
+${NDK_HOME}/build/tools/make_standalone_toolchain.py --unified-headers --api 14 --arch x86 --install-dir NDK/x86
+```
+
+Create a new file, `cargo-config.toml`. This file will tell cargo where to look for the NDKs during cross compilation. Add the following content to the file, remembering to replace instances of `<project path>` with the path to your project directory
+```
+[target.aarch64-linux-android]
+ar = "<project path>/greetings/NDK/arm64/bin/aarch64-linux-android-ar"
+linker = "<project path>/greetings/NDK/arm64/bin/aarch64-linux-android-clang"
+
+[target.armv7-linux-androideabi]
+ar = "<project path>/greetings/NDK/arm/bin/arm-linux-androideabi-ar"
+linker = "<project path>/greetings/NDK/arm/bin/arm-linux-androideabi-clang"
+
+[target.i686-linux-android]
+ar = "<project path>/greetings/NDK/x86/bin/i686-linux-android-ar"
+linker = "<project path>/greetings/NDK/x86/bin/i686-linux-android-clang"
+```
+
+We can leave this here if we only want to compile for Android this one time, however if this is something we think we'll want to do several times, we should copy this file to Cargo's config directory.
+```
+cp cargo-config.toml ~/.cargo/config
+```
+
+Let's go ahead and add our newly created Android architectures to rustup so we can use them during cross compilation:
+```
+rustup target add aarch64-linux-android armv7-linux-androideabi i686-linux-android
+```
+
+Now we're all setup and we're ready to start. Let's create the rust directory. If you've already created a Rust project from following the iOS post, you don't need to do it again.
```

Comment:
s/setup/set up

Can we pick a different name other than "rust"? That seems pretty confusing to me.

Reply:
We did `cargo new cargo` in the iOS post

Reply:
I changed it to "rust" because two people felt that the name "cargo" in the iOS post was inconsistent with the "ios" directory name we were using.